frame/staking-rewards/src/lib.rs

```diff
@@ -372,6 +388,61 @@ pub mod pallet {
 			<Self as Staking>::split(&who, &position, ratio.value())?;
 			Ok(())
 		}
+
+		/// Update a new reward pool based on the provided amounts and rates.
+		///
+		/// Emits `RewardPoolUpdated` event when successful.
+		#[pallet::weight(T::WeightInfo::update_reward_pool())]
+		#[transactional]
+		pub fn update_reward_pool(
+			origin: OriginFor<T>,
+			pool_id: T::RewardPoolId,
+			reward_updates: RewardUpdatesOf<T>,
+		) -> DispatchResult {
+			T::RewardPoolUpdateOrigin::ensure_origin(origin)?;
+
+			let current_time = T::UnixTime::now().as_secs();
+
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let mut inner_rewards = rewards_pool.rewards.into_inner();
+
+			let elapsed_time = current_time - rewards_pool.last_updated;
+
+			for (asset_id, reward) in inner_rewards.iter_mut() {
+				match reward_updates.get(asset_id) {
+					Some(reward_update) => {
+						ensure!(
+							reward.max_rewards > reward.total_rewards + reward_update.amount,
+							Error::<T>::MaxRewardExceeded
+						);
+
+						reward.total_rewards = reward
+							.total_rewards
+							.safe_add(&reward_update.amount)
+							.map_err(|_| ArithmeticError::Overflow)?;
+						// reward.total_rewards = reward.total_rewards
+						// 	.safe_add(elapsed_time * pool.reward_rate)
+						// 	.map_err(|_| ArithmeticError::Overflow)?;
+						reward.reward_rate = reward_update.reward_rate;
+					},
+					None => (),
+				}
+			}
```

Comment:
As a note, this is possible without mutability. It would be much nicer if it was possible to iterate over a bounded map directly - I've opened https://github.com/paritytech/substrate/issues/11866 to suggest an API to provide that functionality

Reply:
That would be much nicer to read!